### PR TITLE
Ability to test with Pre-release versions

### PR DESCRIPTION
**Pull request by stopthatastronaut**
Does what it says on the tin

### CODE REVIEW

File: build-azure.ps1 [hunk 1]
```diff
@@ -24,19 +33,22 @@ Test-CustomVersionOfVagrantDscPluginIsInstalled
...
+    $result = Invoke-Pester -OutputFile PesterTestResults.xml -OutputFormat NUnitXml -PassThru
+    if ($result.FailedCount -gt 0) {
+      exit 1
+    }
 }
 
 Write-Output "Running 'vagrant up --provider azure'"
-vagrant up --provider azure # --debug &> vagrant.log
+
+Invoke-VagrantWithRetries -provider azure
```
[the remark this answers is not in the document]
Comment (stopthatastronaut) [reply build-azure.ps1]: at the moment, they're params on `Invoke-VagrantWithRetries`, but I see your point. I considered that, and I figured they'd be relatively rarely used? Or maybe not.

File: Tests/powershell-helpers.ps1 [hunk 2]
```diff
@@ -59,3 +59,121 @@ function Test-CustomVersionOfVagrantDscPluginIsInstalled() {  # does not deal we
...
+  if($retainondestroy) {
+    $args += '--no-destroy-on-error'
+  }
+
+  if($debug) {
+    $args += '--debug'
+  }
+
+  do {
+    Write-Output "Running 'vagrant up --provider $provider"
```
[the remark this answers is not in the document]
Comment (stopthatastronaut) [reply Tests/powershell-helpers.ps1]: Good call. On it.

File: Tests/powershell-helpers.ps1 [hunk 3]
```diff
@@ -59,3 +59,121 @@ function Test-CustomVersionOfVagrantDscPluginIsInstalled() {  # does not deal we
...
+  } while ($retryAgain)
+}
+
+Function Set-OctopusDSCEnvVars {
+  param(
+    [switch]$offline,
+    [switch]$SkipPester,
+    [switch]$ServerOnly,
+    [switch]$TentacleOnly,
+    [string]$PreReleaseVersion
```
[the remark this answers is not in the document]
Comment (stopthatastronaut) [reply Tests/powershell-helpers.ps1]: As long as it's a version in the `octopus-testing` bucket, but yeah, good point.

[the remark this answers is not in the document]
Comment (stopthatastronaut) [reply Tests/powershell-helpers.ps1]: Given that we test upgrading, should we be able to specify the UpgradeFromVersion and the... errr... "Main" version?

[the remark this answers is not in the document]
Comment (matt-richardson) [reply Tests/powershell-helpers.ps1]: Not sure that’s necessary at his point. Maybe if we find we need it in future?

[the remark this answers is not in the document]
Comment (stopthatastronaut) [reply Tests/powershell-helpers.ps1]: OK. Swapped to `$OctopusVersion`

File: build-azure.ps1 (hunk 1 above)
[the remark this answers is not in the document]
Comment (stopthatastronaut) [reply build-azure.ps1]: Done

File: Tests/surface-logs.ps1 [hunk 4]
```diff
@@ -1,21 +1,26 @@
...
+                $ConfigPaths += $Config.ConfigurationFilePath
+            }
         }
     }
 
     if( (Test-Path "HKLM:\SOFTWARE\Octopus\OctopusServer\"))
     {
         Get-ChildItem  "HKLM:\SOFTWARE\Octopus\OctopusServer\" | ForEach-Object {
             $_ | ForEach-Object {
+                write-host "adding '$(Get-ItemProperty $_.PSPath | select -expand Configurationfilepath)'"
```
[the remark this answers is not in the document]
Comment (stopthatastronaut) [reply Tests/surface-logs.ps1]: I'll get back to that one.

**Pull request merged**